new scripts for Authentik

remz1337 opened the pull request:
> [!NOTE]
> We are meticulous when it comes to merging code into the main branch, so please understand that we may reject pull requests that do not meet the project's standards. It's never personal. Also, game-related scripts have a lower chance of being merged.

## Description

Adding scripts for [Authentik](https://github.com/goauthentik/authentik), as requested in this discussion (https://github.com/community-scripts/ProxmoxVE/discussions/53)

Installation has been tested working by a few users, but update script is yet to be tested.

Installation takes about 30 minutes to build everything. If you see any potential optimizations, let me know.

## Type of change
Please check the relevant option(s):

- [ ] Bug fix (non-breaking change that resolves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (a fix or feature that would cause existing functionality to change unexpectedly)
- [X] New script (a fully functional and thoroughly tested script or set of scripts.)

## Prerequisites
The following efforts must be made for the PR to be considered. Please check when completed:
- [X] Self-review performed (I have reviewed my code, ensuring it follows established patterns and conventions)
- [X] Testing performed (I have tested my changes, ensuring everything works as expected)
- [X] Documentation updated (I have updated any relevant documentation)

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,185 @@
+$STD update-alternatives --install /usr/bin/python3 python3 /usr/local/bin/python3.12 1
+msg_ok "Installed Python 3.12"
+
+NODE_VER="22"
```
Does it have to be the 22, or can we also use the standard NodeJS from Debian? In the end, that was only 3 lines instead of 8 or 9

remz1337 replied:
Might be worth a try, but this step either it be 4 or 8 line doesn't take more/less time. NodeJS v22 is what Authentik uses in it's Docker image so I think it's best to stick with the same dependencies and avoid potential compatibility issues

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,185 @@
+echo "${RELEASE}" >/opt/${APPLICATION}_version.txt
+msg_ok "Built Authentik website"
+
+msg_info "Installing Golang"
```
Can we Switch GoLang above the authentik Installation Part?

remz1337 replied:
yes, makes sense

MickLesk reviewed (commented):
First little review, more later

havardthom commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,202 @@
+Environment=DJANGO_SETTINGS_MODULE="authentik.root.settings"
+ExecStart=celery -A authentik.root.celery worker -Ofair --max-tasks-per-child=1 --autoscale 3,1 -E -B -s /tmp/celerybeat-schedule -Q authentik,authentik_scheduled,authentik_events
+WorkingDirectory=/opt/authentik/authentik
+#User=authentik
```
remove comments

havardthom commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,202 @@
+#Group=authentik
+Restart=always
+RestartSec=5
+[Install]
```
```suggestion

[Install]
```

havardthom commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,202 @@
+WantedBy=multi-user.target
+EOF
+systemctl enable -q --now authentik-worker
+sleep 5
```
```suggestion
```
is this needed?

havardthom commented on json/authentik.json:
```diff
@@ -0,0 +1,39 @@
+    ],
+    "date_created": "2024-11-06",
+    "type": "ct",
+    "updateable": false,
```
isn't it updateable?

remz1337 replied:
yes! thanks

havardthom commented on json/authentik.json:
```diff
@@ -0,0 +1,39 @@
+    "updateable": false,
+    "privileged": false,
+    "interface_port": "9000",
+    "documentation": null,
```
```suggestion
    "documentation": "https://docs.goauthentik.io/docs/",
```

havardthom commented on json/authentik.json:
```diff
@@ -0,0 +1,39 @@
+    "notes": [
+        {
+            "text": "Initial configuration at http://<IP>:9000/if/flow/initial-setup/",
+            "type": "warning"
```
```suggestion
            "type": "info"
```

MickLesk commented:
Ive do some changes (locally), tomorrow i test it on a more powerful device, then my Dev device

newzealandpaul commented:
@remz1337 can you commit @havardthom suggestions if you agree with them, or dismiss otherwise.

remz1337 commented on ct/authentik.sh:
```diff
@@ -0,0 +1,112 @@
+if [[ ! -f /etc/systemd/system/authentik-server.service ]]; then msg_error "No ${APP} Installation Found!"; exit; fi
+RELEASE=$(curl -s https://api.github.com/repos/goauthentik/authentik/releases/latest | grep "tarball_url" | awk '{print substr($2, 2, length($2)-3)}')
+if [[ "${RELEASE}" != "$(cat /opt/${APP}_version.txt)" ]] || [[ ! -f /opt/${APP}_version.txt ]]; then
+  whiptail --backtitle "Proxmox VE Helper Scripts" --msgbox --title "SET RESOURCES" "Please set the resources in your ${APP} LXC to ${var_cpu}vCPU and ${var_ram}RAM for the build process before continuing" 10 75
```
good catch, I'll remove it

remz1337 commented:
Thanks for your patience. I've been very busy lately. I've addressed the feedback given. Could you do another review? @MickLesk  @havardthom @newzealandpaul

havardthom commented on ct/authentik.sh:
```diff
@@ -0,0 +1,111 @@
+  tar -xzf authentik.tar.gz -C /opt/authentik --strip-components 1 --overwrite
+  rm -rf authentik.tar.gz
+  cd /opt/authentik/website
+  npm install >/dev/null 2>&1
```
```suggestion
  npm install &>/dev/null
```
Replace everywhere

havardthom commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,202 @@
+msg_ok "Installed Python 3.12"
+
+NODE_VER="22"
+msg_info "Installing Node.js ${NODE_VER}"
```
```suggestion
msg_info "Setting up Node.js Repository"
mkdir -p /etc/apt/keyrings
curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg
echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_22.x nodistro main" >/etc/apt/sources.list.d/nodesource.list
msg_ok "Set up Node.js Repository"

msg_info "Installing Node.js"
$STD apt-get update
$STD apt-get install -y nodejs
msg_ok "Installed Node.js"
```
Remove environment variable and split the section like other scripts

havardthom reviewed (commented):
There are still multiple comments from my previous review that have not been addressed, you need to expand conversations

![image](https://github.com/user-attachments/assets/2b02f56a-ecf9-4ae8-af0f-8d53fac0c171)

MickLesk commented:
@remz1337 any Update?

remz1337 commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,202 @@
+mkdir -p /etc/authentik
+cp /opt/authentik/authentik/lib/default.yml /opt/authentik/authentik/lib/default.yml.BAK
+mv /opt/authentik/authentik/lib/default.yml /etc/authentik/config.yml
+$STD yq -i ".secret_key = \"$(openssl rand -hex 32)\"" /etc/authentik/config.yml
```
true, but I think managing yml with yq is easier and cleaner. But given yq is only needed during installation, I have added a line to remove it at the end of the installation

remz1337 commented:
@MickLesk @havardthom thanks for your patience. I've addressed the comments. Please let me know if there's anything else

MickLesk commented:
Script breaking at my try

![image](https://github.com/user-attachments/assets/0bf51deb-68bb-4b77-88bf-9a34a379cecc)

remz1337 commented:
@MickLesk there was an issue with my `build.func`. Fixed now. I have successfully ran the installation. Can you try again? Thanks

MickLesk commented:
Test tomorrow @work. My dev nuc is too weak for the script 😂

MickLesk commented:
Script work for me, but some improvements:
1. script does not comply with our standards for the most part, would have to be adapted across the board.

2. the boot disk is a bit small, I would probably even say 15GB here, otherwise the resource check will pop in with every update.

I'll go through the script and write some remarks on it

<img width="1540" alt="image" src="https://github.com/user-attachments/assets/a63a4352-0f1a-4562-ab02-1056c52ef677">

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+update_os
+
+msg_info "Installing Dependencies (Patience)"
+$STD apt-get install -y --no-install-recommends \
```
--no-install-recommends makes sense? We don't really use it anywhere anymore, as it had caused problems more often

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,185 @@
+chmod +x /usr/bin/yq
+msg_ok "Installed yq"
+
+msg_info "Installing Python 3.12"
```
then switch the header text to "Setup Python3" and msg_ok "Setup Python3"

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+wget -q https://golang.org/dl/${GO_RELEASE}
+tar -xzf ${GO_RELEASE} -C /usr/local
+ln -s /usr/local/go/bin/go /usr/bin/go
+rm -rf go/
+rm -rf ${GO_RELEASE}
```
should be at the bottom in the cleaning up area

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+$STD go build -o /opt/authentik/authentik-server /opt/authentik/cmd/server/
+msg_ok "Built Go Proxy"
+
+msg_info "Installing GeoIP"
+cd ~
+GEOIP_RELEASE=$(curl -s https://api.github.com/repos/maxmind/geoipupdate/releases/latest | grep "tag_name" | awk '{print substr($2, 3, length($2)-4) }')
+wget -qO geoipupdate.deb https://github.com/maxmind/geoipupdate/releases/download/v${GEOIP_RELEASE}/geoipupdate_${GEOIP_RELEASE}_linux_amd64.deb
+$STD dpkg -i geoipupdate.deb
+rm geoipupdate.deb
+cat <<EOF >/etc/GeoIP.conf
+#GEOIPUPDATE_EDITION_IDS="GeoLite2-City GeoLite2-ASN"
+#GEOIPUPDATE_VERBOSE="1"
+#GEOIPUPDATE_ACCOUNT_ID_FILE="/run/secrets/GEOIPUPDATE_ACCOUNT_ID"
+#GEOIPUPDATE_LICENSE_KEY_FILE="/run/secrets/GEOIPUPDATE_LICENSE_KEY"
+EOF
+msg_ok "Installed GeoIP"
```
After NodeJS Install? Its an dependency and not the main prog

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+set -o pipefail
+msg_ok "Installed Golang"
+
+msg_info "Building authentik website"
```
Setup Authentik

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+mkdir -p /opt/authentik
+wget -qO authentik.tar.gz "${RELEASE}"
+tar -xzf authentik.tar.gz -C /opt/authentik --strip-components 1 --overwrite
+rm -rf authentik.tar.gz
```
in cleaning up section

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+echo "${RELEASE}" >/opt/${APPLICATION}_version.txt
+msg_ok "Built authentik website"
+
+msg_info "Building Go Proxy"
```
all in msg_info "Setup Authentik"

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+$STD pip install .
+msg_ok "Installed Python Dependencies"
+
+msg_info "Installing Redis"
+$STD apt-get install -y redis-server
+systemctl enable -q --now redis-server
+msg_ok "Installed Redis"
```
move above after NodeJS Installing

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+systemctl enable -q --now redis-server
+msg_ok "Installed Redis"
+
+msg_info "Installing PostgreSQL"
+$STD apt-get install -y postgresql postgresql-contrib
+DB_NAME="authentik"
+DB_USER="authentik"
+DB_PASS="$(openssl rand -base64 18 | cut -c1-13)"
+$STD sudo -u postgres psql -c "CREATE DATABASE $DB_NAME;"
+$STD sudo -u postgres psql -c "CREATE USER $DB_USER WITH PASSWORD '$DB_PASS';"
+$STD sudo -u postgres psql -c "GRANT ALL PRIVILEGES ON DATABASE $DB_NAME TO $DB_USER;" 
+$STD sudo -u postgres psql -c "ALTER DATABASE $DB_NAME OWNER TO $DB_USER;"
+$STD sudo -u postgres psql -c "ALTER USER $DB_USER WITH SUPERUSER;"
+msg_ok "Installed PostgreSQL"
```
move above after NodeJS Installing

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+$STD bash /opt/authentik/lifecycle/ak migrate
+msg_ok "Installed authentik"
+
+msg_info "Configuring Services"
```
msg_info "Creating Service"

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+[Install]
+WantedBy=multi-user.target
+EOF
+systemctl enable -q --now authentik-server
+sleep 2
```
move below, before systemctl enable -q --now authentik-worker

MickLesk commented on install/authentik-install.sh:
```diff
@@ -0,0 +1,201 @@
+motd_ssh
+customize
+
+msg_info "Cleaning up"
```
Add here all cleaning rm's

MickLesk commented on json/authentik.json:
```diff
@@ -0,0 +1,39 @@
+    },
+    "notes": [
+        {
+            "text": "Initial configuration at http://<IP>:9000/if/flow/initial-setup/",
+            "type": "info"
```
not really needed, is displayed at the end of the installation process anyway

MickLesk reviewed (commented):
i think all "authentik" Install parts should be 1 Part:
msg_info "Setup Authentik"
...
...
...
msg_ok "Setup Authentik" 

i mean "Building authentik website", then "Building Go Proxy" and "Installing authentik" in one part. (incl. the Installing Python Dependencies"

MickLesk commented:
Wild outputs.
Can you drag the CT and install.sh onto the develop branch so that it can be tested? 

<img width="528" alt="image" src="https://github.com/user-attachments/assets/aee80526-beee-4c47-ad6f-e211e718df6b" />

MickLesk commented:
@community-scripts/contributor 

Can be tested here:
```bash
bash -c "$(wget -qLO - https://github.com/community-scripts/ProxmoxVE/raw/develop/ct/authentik.sh)"
```

lasharor commented:
![image](https://github.com/user-attachments/assets/b028b826-82a0-430e-95b6-f946d3691fe9)

MickLesk commented:
can you try again in 10 min after github refresh my latest changes? @lasharor 

my dev nuc breaks at this script :D (to less cpu's)

michelroegl-brunner commented:
Installs fine on my machine.

lasharor commented:
took some time to install but now also works fine. 

At the end the script provides the IP of the LXC. But doesn't give the full path for the config flow. This was present in previous versions of this script. I think it is better to display IP:9000/if/flow/initial-setup/

remz1337 commented:
Thanks @MickLesk  for pulling it into the dev branch. Should we open a new PR to merge from dev and close this one?

MickLesk commented:
i think we fix a little things in dev and update this pr files after this. then we can approve and merge

second2falcon commented:
> took some time to install but now also works fine.
> 
> At the end the script provides the IP of the LXC. But doesn't give the full path for the config flow. This was present in previous versions of this script. I think it is better to display IP:9000/if/flow/initial-setup/

This indeed is neccesary. If I wouldn't have known that Authentik has a specific path and port during setup, I might've given up.

DJKatastrof commented:
This been loading for 20 minutes here, is it normal?
![image](https://github.com/user-attachments/assets/777ccd7e-163c-4933-b416-2431af992d73)
